Collect additional BTRFS metrics

Dim-P opened the pull request:
##### Summary
This PR extends the BTRFS metrics we collect. 

Closes https://github.com/netdata/netdata/issues/3341, closes https://github.com/netdata/netdata/issues/13561, closes https://github.com/netdata/netdata/issues/14494 .

##### TODO:
- [x] `/sys/fs/btrfs/[UUID]/commit_stats`
Metrics related to commit stats. Two new charts (per BTRFS filesystem) have been added, one showing commits rate (1 dimension) and another one showing timings metrics (3 dimensions):

![Screenshot from 2023-03-03 16-44-14](https://user-images.githubusercontent.com/5953192/222777836-989b21a4-873c-4a4c-8e48-c90a75501e8c.png)


- [x] `/sys/fs/btrfs/[UUID]/devinfo/[DEVID]/error_stats`
Metrics related to device errors. One new chart has been added (per BTRFS filesystem). One dimension is created per error metric per device, prefixed with the device ID:

![Screenshot from 2023-03-03 16-44-57](https://user-images.githubusercontent.com/5953192/222778005-b12229c3-f924-4251-b4fc-11158111d6e1.png)


- [ ] `/sys/fs/btrfs/[UUID]/devinfo/[DEVID]/missing`
Won't be added for now, as it doesn't seem to behave as expected (see https://github.com/netdata/netdata/issues/14494#issuecomment-1452163022)
- [ ] `/sys/fs/btrfs/[UUID]/devinfo/[DEVID]/replace_target`  
Won't be implemented as part of this PR.

##### Test Plan

Install this branch and check if the expected BTRFS metrics are reported for all your BTRFS filesystems and if they are the same as the `/sys/fs/btrfs/*` stats.

##### Additional Information
<!-- This is usually used to help others understand your
motivation behind this change. A step-by-step reproduction of the problem is
helpful if there is no related issue. -->

<details> <summary>For users: How does this change affect me?</summary>
New charts and alarms for BTRFS filesystem stats that will help in debugging errors or preventing filesystem failures. 
</details>

Dim-P commented:
@Ferroin Feel free to test the current state of the PR and if you are happy with the charts I have added, please add descriptions and alerts to them.

Ferroin commented:
Mostly looks good, but there are three things I think we should ideally change here:

- The commit chart is likely to be much more useful if it’s tracking the rate of commits instead of the total number. That would allow users (and alarms) to more easily see when commits are happening, which is the most useful piece of information this chart can provide.
- Similarly, the total commit time should probably be handled incrementally so that users can see how much time is being spent on commits when. The other two commit timing dimensions largely make sense as they are now (functioning essentially as peak detectors)
- For the error rates, the chart and all dimensions should exist even if there are no errors reported. Whether or not it gets displayed is of course up to the dashboard itself, but it needs to exist so that it we can have proper alarms show up for it in the list of configured alarms even if there are no errors (and so that anything scraping the API actually sees it).

Irrespective of those suggestions, I’ll look at getting chart descriptions and a default alarm config for the error counts added either later today or some time Monday.

Ferroin commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -736,6 +1003,172 @@ int do_sys_fs_btrfs(int update_every, usec_t dt) {
+                    }
+                    rrddim_set_by_pointer(node->st_error_stats, d->rd_read_errs, d->read_errs);
+                }
+                if(d->flush_errs){
+                    if(unlikely(!d->rd_flush_errs)){
+                        snprintfz(rd_id, RRD_ID_LENGTH_MAX, "%d:flush_errs", d->id);
+                        d->rd_flush_errs = rrddim_add(node->st_error_stats, rd_id, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                    }
+                    rrddim_set_by_pointer(node->st_error_stats, d->rd_flush_errs, d->flush_errs);
+                }
+                if(d->flush_errs){
+                    if(unlikely(!d->rd_flush_errs)){
+                        snprintfz(rd_id, RRD_ID_LENGTH_MAX, "%d:flush_errs", d->id);
+                        d->rd_flush_errs = rrddim_add(node->st_error_stats, rd_id, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                    }
+                    rrddim_set_by_pointer(node->st_error_stats, d->rd_flush_errs, d->flush_errs);
+                }
```
Possible duplicate code?

Ferroin commented:
@Dim-P basic alerts and chart descriptions added. I think this is probably ready for review.

Dim-P commented:
> Mostly looks good, but there are three things I think we should ideally change here:
> 
>     * The commit chart is likely to be much more useful if it’s tracking the rate of commits instead of the total number. That would allow users (and alarms) to more easily see when commits are happening, which is the most useful piece of information this chart can provide.
> 
>     * Similarly, the total commit time should probably be handled incrementally so that users can see how much time is being spent on commits when. The other two commit timing dimensions largely make sense as they are now (functioning essentially as peak detectors)
> 
>     * For the error rates, the chart and all dimensions should exist even if there are no errors reported. Whether or not it gets displayed is of course up to the dashboard itself, but it needs to exist so that it we can have proper alarms show up for it in the list of configured alarms even if there are no errors (and so that anything scraping the API actually sees it).
> 
> 
> Irrespective of those suggestions, I’ll look at getting chart descriptions and a default alarm config for the error counts added either later today or some time Monday.

I've made the changes, I'll open it up for review.

ilyam8 commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -736,6 +997,147 @@ int do_sys_fs_btrfs(int update_every, usec_t dt) {
+                        , RRDSET_TYPE_LINE
+                );
+
+                node->rd_commit_timings_last = rrddim_add(node->st_commit_timings, "last commit", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                node->rd_commit_timings_max = rrddim_add(node->st_commit_timings, "max commit", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                node->rd_commit_timings_total = rrddim_add(node->st_commit_timings, "current commit", NULL, 1, 1, RRD_ALGORITHM_INCREMENTAL);
```
Please split this into two charts: don't mix absolute and incremental dimensions.

Dim-P replied:
Done.

ilyam8 replied:
Why? Looks like 2 different metrics to me:
 - (last, max) commit time
 - commit**s** time.

thiagoftsm replied:
Hello @Dim-P I agree with you that we should collect anyway, the unique concern here is to group different kind of metrics in the same chart.

Dim-P replied:
Hey @thiagoftsm , we did discuss this with @ilyam8 and @Ferroin and decided to create a new chart that shows the commits time share as a percentage value. See latest commit and let me know what you think.

thiagoftsm commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -736,6 +997,147 @@ int do_sys_fs_btrfs(int update_every, usec_t dt) {
+            if(unlikely(!node->st_commits)) {
+                char id[RRD_ID_LENGTH_MAX + 1], name[RRD_ID_LENGTH_MAX + 1], title[200 + 1];
+
+                snprintf(id, RRD_ID_LENGTH_MAX, "commits_%s", node->id);
```
Please, change calls to [snprintfz](https://github.com/netdata/netdata/blob/af360d8b0f31d6fe463928a0f26787861daaa4c2/libnetdata/libnetdata.h#L387).

Dim-P replied:
The rest of the code in that source file uses `snprintf()`, so I just used the same function for consistency. It doesn't make a difference in any case, since the return value is not checked and effectively `snprintf()` and `snprintfz()` have the same behavior in this case.

thiagoftsm replied:
Ok, considering that the rest of the code is using this pattern, let us change this with  other PR to have only one pattern in our code.

thiagoftsm commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -736,6 +997,147 @@ int do_sys_fs_btrfs(int update_every, usec_t dt) {
+            if(unlikely(!node->st_error_stats)) {
+                char id[RRD_ID_LENGTH_MAX + 1], name[RRD_ID_LENGTH_MAX + 1], title[200 + 1];
+
+                snprintf(id, RRD_ID_LENGTH_MAX, "error_stats_%s", node->id);
```
The same here.

Dim-P commented:
@ilyam8 @Ferroin @thiagoftsm  Can you please check the charts after my latest commit? 

I added a new chart that show the commits time share as a percentage value and I also changed the commits chart to use an absolute dimension, as the incremental one didn't align well with all the other metrics (timings etc.) being absolute, due to interpolation. I thought that would confuse users, since it was the only one shown as an interpolated rate.

I also updated the descriptions, so any feedback is welcome.

ilyam8 commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -736,6 +1020,178 @@ int do_sys_fs_btrfs(int update_every, usec_t dt) {
+
+
+            for(BTRFS_DEVICE *d = node->devices ; d ; d = d->next) {
+                char rd_id[RRD_ID_LENGTH_MAX + 1];
+                if(unlikely(!d->rd_write_errs)){
+                    snprintfz(rd_id, RRD_ID_LENGTH_MAX, "%d:write_errs", d->id);
+                    d->rd_write_errs = rrddim_add(node->st_error_stats, rd_id, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                    snprintfz(rd_id, RRD_ID_LENGTH_MAX, "%d:read_errs", d->id);
+                    d->rd_read_errs = rrddim_add(node->st_error_stats, rd_id, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                    snprintfz(rd_id, RRD_ID_LENGTH_MAX, "%d:flush_errs", d->id);
+                    d->rd_flush_errs = rrddim_add(node->st_error_stats, rd_id, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                    snprintfz(rd_id, RRD_ID_LENGTH_MAX, "%d:corruption_errs", d->id);
+                    d->rd_corruption_errs = rrddim_add(node->st_error_stats, rd_id, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                    snprintfz(rd_id, RRD_ID_LENGTH_MAX, "%d:generation_errs", d->id);
```
I think you need to create a chart per device, not a set of dimensions per device. If you do it like this these charts will be perfectly aggregated.

Ferroin replied:
Hmm, good point about aggregation, one chart per device is probably the best option here.

ilyam8 replied:
Yeah, if we keep it as it is - the charts are not aggregatable.

ilyam8 replied:
I can only repeat that dimension "per device per X (read/write errors in your case)" is wrong, should be a chart per device (if I understand correctly, and `1` and `2` are devices).

Dim-P replied:
I stand corrected, I had misunderstood how grouping works on the cloud (and that error charts can be grouped by context in this particular case). In which case what you said makes total sense. 

I pushed a new commit addressing this.

Ferroin commented:
> as the incremental one didn't align well with all the other metrics (timings etc.) being absolute, due to interpolation. I thought that would confuse users, since it was the only one shown as an interpolated rate.

I see three problems with this being an absolute dimension:

- The absolute number of commits pretty much _never_ matters unless you’re doing highly specialized testing of the filesystem itself and need to confirm that at least one commit happened since you last looked. For 99.99% of users, what actually matters is the rate of commits, and even then mostly just as a way to explain some parts of disk usage patterns.
- For filesystems that have been mounted for a very long time, checking that commits have been happening consistently is much easier to do with an incremental dimension than an absolute dimension (instead of having to just check for a flat line, you have to check the slope and try to estimate how consistent it is).
- An absolute dimension for this will make interpreting aggregate data more difficult, especially when dealing with filesystems that are not mounted all the time.

I also kind of doubt that users will get confused as long as we properly explain that it’s the commit rate in commits per second, but that’s a bit more debatable.

Dim-P commented:
> > as the incremental one didn't align well with all the other metrics (timings etc.) being absolute, due to interpolation. I thought that would confuse users, since it was the only one shown as an interpolated rate.
> 
> I see three problems with this being an absolute dimension:
> 
>     * The absolute number of commits pretty much _never_ matters unless you’re doing highly specialized testing of the filesystem itself and need to confirm that at least one commit happened since you last looked. For 99.99% of users, what actually matters is the rate of commits, and even then mostly just as a way to explain some parts of disk usage patterns.
> 
>     * For filesystems that have been mounted for a very long time, checking that commits have been happening consistently is much easier to do with an incremental dimension than an absolute dimension (instead of having to just check for a flat line, you have to check the slope and try to estimate how consistent it is).
> 
>     * An absolute dimension for this will make interpreting aggregate data more difficult, especially when dealing with filesystems that are not mounted all the time.
> 
> 
> I also kind of doubt that users will get confused as long as we properly explain that it’s the commit rate in commits per second, but that’s a bit more debatable.

To be clear, the results are similar to the first chart seen in https://user-images.githubusercontent.com/5953192/222777836-989b21a4-873c-4a4c-8e48-c90a75501e8c.png , but instead of a single commit being interpolated and spanning 2 collection intervals (e.g. as 0.64 commits on first collection interval and 0.36 commits on the next one), it will only show in the collection interval it happened (e.g. as 1.0 commit on just one interval). 

It's not a flat line, it is still commit spikes when they happen, but without interpolation.

ilyam8 commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -449,16 +703,18 @@ static inline int find_all_btrfs_pools(const char *path) {
 static void add_labels_to_btrfs(BTRFS_NODE *n, RRDSET *st) {
-    rrdlabels_add(st->rrdlabels, "device", n->id, RRDLABEL_SRC_AUTO);
-    rrdlabels_add(st->rrdlabels, "device_label", n->label, RRDLABEL_SRC_AUTO);
+    rrdlabels_add(st->rrdlabels, "filesystem", n->id, RRDLABEL_SRC_AUTO);
```
```suggestion
    rrdlabels_add(st->rrdlabels, "filesystem_uuid", n->id, RRDLABEL_SRC_AUTO);
```

or `id`

ilyam8 commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -736,6 +1019,177 @@ int do_sys_fs_btrfs(int update_every, usec_t dt) {
+                            , id
+                            , name
+                            , node->label
+                            , "btrfs.errors"
```
```suggestion
                            , "btrfs.device_errors"
```

Because it is device errors

ilyam8 commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -736,6 +1019,177 @@ int do_sys_fs_btrfs(int update_every, usec_t dt) {
+                            , RRDSET_TYPE_LINE
+                    );
+
+                    char rd_id[RRD_ID_LENGTH_MAX + 1];
+                    snprintfz(rd_id, RRD_ID_LENGTH_MAX, "write_errs");
+                    d->rd_write_errs = rrddim_add(d->st_error_stats, rd_id, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+                    snprintfz(rd_id, RRD_ID_LENGTH_MAX, "read_errs");
```
Why absolute? Should we show error rate (incremental)?

Dim-P replied:
As errors indicate corruption or hardware error, I believe users would want to keep them to 0. Once they have fixed the error by repairing the filesystem or replacing a device, they can reset the error counters with something like `btrfs device stats --reset` . @Ferroin what's your opinion?

ilyam8 commented on collectors/proc.plugin/sys_fs_btrfs.c:
```diff
@@ -736,6 +1019,177 @@ int do_sys_fs_btrfs(int update_every, usec_t dt) {
+                    snprintfz(rd_id, RRD_ID_LENGTH_MAX, "generation_errs");
+                    d->rd_generation_errs = rrddim_add(d->st_error_stats, rd_id, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+
+                    add_labels_to_btrfs(node, d->st_error_stats);
```
Device error should have a `device_id` label too.

ilyam8 replied:
It is not the same because `instance` is the chart id (the whole thing with filesystem uuid).

Dim-P commented:
@ilyam8 @Ferroin Happy with this PR as it is now? Any more requests for changes?

Ferroin commented:
> @ilyam8 @Ferroin Happy with this PR as it is now? Any more requests for changes?

The `lookup` lines in the alarm config need updated to sync up with the dimension name changes, but otherwise LGTM.

Dim-P commented:
Alarm configuration updated and tested, should be ready to go:

![Screenshot from 2023-04-06 18-41-15](https://user-images.githubusercontent.com/5953192/230457604-90e7c772-6435-4ce9-9e07-4aeb7481497a.png)

ilyam8 commented:
No BTRFS metrics after installing this PR. What I did:

 - setup BTRFS and install this branch

```bash
dd if=/dev/zero of=/loopbackfile1.img bs=100M count=10
dd if=/dev/zero of=/loopbackfile2.img bs=100M count=10

losetup  /dev/loop1 /loopbackfile1.img
losetup  /dev/loop2 /loopbackfile2.img

mkfs.btrfs /dev/loop1
mkfs.btrfs /dev/loop2

mkdir /opt/btrfsTest1
mkdir /opt/btrfsTest2

mount /dev/loop1 /opt/btrfsTest1
mount /dev/loop2 /opt/btrfsTest2
```

---

- Logs

```
collector.log:2023-04-06 22:13:48: netdata ERROR : P[proc] : BTRFS: failed to read '/sys/fs/btrfs/d39812bf-2782-47d9-90af-7ba2b0da5fca/commit_stats' (errno 2, No such file or directory)
collector.log:2023-04-06 22:13:48: netdata ERROR : P[proc] : BTRFS: failed to read '/sys/fs/btrfs/e11d2f87-65d0-44da-82a3-d3725ba7921e/commit_stats' (errno 2, No such file or directory)
```

- `ls -l`

```bash
$ ls -l /sys/fs/btrfs/d39812bf-2782-47d9-90af-7ba2b0da5fca/
drwxr-xr-x    - root  6 Apr 22:07 allocation
drwxr-xr-x    - root  6 Apr 22:07 devices
drwxr-xr-x    - root  6 Apr 22:14 devinfo
drwxr-xr-x    - root  6 Apr 22:14 features
lrwxrwxrwx    0 root  6 Apr 22:14 bdi -> ../../../devices/virtual/bdi/btrfs-1
.r--r--r-- 4.1k root  6 Apr 22:14 checksum
.r--r--r-- 4.1k root  6 Apr 22:14 clone_alignment
.r--r--r-- 4.1k root  6 Apr 22:14 exclusive_operation
.rw-r--r-- 4.1k root  6 Apr 22:07 label
.r--r--r-- 4.1k root  6 Apr 22:14 metadata_uuid
.r--r--r-- 4.1k root  6 Apr 22:14 nodesize
.rw-r--r-- 4.1k root  6 Apr 22:14 quota_override
.r--r--r-- 4.1k root  6 Apr 22:14 sectorsize
```

---

I assume lack of `commit_stats` is the problem. I think we should just ignore the file if it doesn't exist, the lack of the file shouldn't result in no BTRFS metrics.

thiagoftsm commented:
@ilyam8 could it have relationship with the btrfs version you are running? I tested latest Linux kernel and I have normal access to metrics.

Btrfs internal functions were modified when kernel `5.10` was released.

thiagoftsm reviewed (approved):
I am seeing metrics as expected, but before to merge, please, take a look in what @ilyam8 described.

Ferroin commented:
> No BTRFS metrics after installing this PR. What I did:
> 
>     * setup BTRFS and install this branch
> 
> 
> Details
> 
> See the logs:
> 
> ```
> collector.log:2023-04-06 22:13:48: netdata ERROR : P[proc] : BTRFS: failed to read '/sys/fs/btrfs/d39812bf-2782-47d9-90af-7ba2b0da5fca/commit_stats' (errno 2, No such file or directory)
> collector.log:2023-04-06 22:13:48: netdata ERROR : P[proc] : BTRFS: failed to read '/sys/fs/btrfs/e11d2f87-65d0-44da-82a3-d3725ba7921e/commit_stats' (errno 2, No such file or directory)
> ```
> 
> Indeed, the file is missing
> ls -l
> 
> I assume the lack of `commit_stats` is the problem. I think we should just ignore the file if it doesn't exist, the lack of the file shouldn't result in no BTRFS metrics.

Correct, it should be logged, but ignored. We also need similar logic actually for the error counters, as older kernels do not have those either.

Dim-P commented:
Implemented the requested changes but discovered a bug when adding and then removing additional devices. Converting it to draft while I am looking into it.

Dim-P commented:
Bug should be fixed now. 

@ilyam8 @Ferroin @thiagoftsm please review again and test particularly the following cases:

1. No commit stats file
2. No error stats files
3. Adding and removing devices onto an existing BTRFS filesystem adds and removes the respective device error charts correctly.

ilyam8 commented on web/gui/dashboard_info.js:
```diff
@@ -6445,6 +6445,22 @@ netdataDashboard.context = {
+    },
+
+    'btrfs.device_errors': {
+        info: 'Tracks per-device error counts. Five types of errors are tracked: read errors, write errors, flush errors, corruption errors, and generation errors. Read, write, and flush errors are errors reported by the underlying block device when trying to perform the associated operations on behalf of BTRFS. Corruption errors count checksum mismatches, which usually are a result of either at- ...
```
```suggestion
        info: 'Tracks per-device error counts. Five types of errors are tracked: read errors, write errors, flush errors, corruption errors, and generation errors. <b>Read</b>, <b>write</b>, and <b>flush</b> are errors reported by the underlying block device when trying to perform the associated operations on behalf of BTRFS. <b>Corruption</b> errors count checksum mismatches, which usually are a result of either at-rest data corruption or hardware problems. <b>Generation</b> errors count generational mismatches within the internal data structures of the volume, and are also usually indicative of at-rest data corruption or hardware problems. Note that errors reported here may not trigger an associated IO error in userspace, as BTRFS has relatively robust error recovery that allows it to return correct data in most multi-device setups.'
```
